internal/controller/apiextensions/claim/reconciler.go

```diff
@@ -559,7 +575,48 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	// We have a watch on both the claim and its composite, so there's no
 	// need to requeue here.
 	cm.SetConditions(xpv1.Available())
-	return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, cm), errUpdateClaimStatus)
+	return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, cm, r.fieldOwner), errUpdateClaimStatus)
+}
+
+// For the background, check https://github.com/kubernetes/kubernetes/issues/99003
+// Managed fields owner key is a pair (the manager name, the used operation).
+// Initially, we create a composite using CREATE request, because that is the only
+// way to figure out if the generated name is already taken.
+// If the operation is successful, managed fields operation will be set to "Update".
+// Later patches are going to produce another managed fields set for the same manager,
+// but with a different operation "Apply".
+// Such configuration would prevent removal of claim fields on the composite, i.e.
+// it would only remove the ownership.
+// In order to fix that, we need to manually change operation to "Apply",
+// before the first composite patch is sent to k8s api server.
+// Returns true if the ownership was fixed.
+func (r *Reconciler) maybeFixFieldOwnership(obj *unstructured2.Unstructured, fixSubresource bool) bool {
```

Comment:
> What exactly do we lose without this migration?

For claims created with the previous version of crossplane, we are going to establish shared field ownership between the new manager and the old one (even when server-side apply is not used, k8s sets the field manager on claim/composite). 

For example, if we create [claim](https://github.com/crossplane/crossplane/blob/44eb7ed980a56b8db8c8a1cb8ba7cf285882f9bf/test/e2e/manifests/apiextensions/composition/propagate-field-removals/claim.yaml) with the latest released Crossplane version and perform:

```sh
kubectl get xnopresources.nop.example.org --show-managed-fields -o yaml
```

K8s API server sers the managed fields:

```yaml
managedFields:
- apiVersion: nop.example.org/v1alpha1
  fieldsType: FieldsV1
  fieldsV1:
    f:metadata:
      f:annotations:
        .: {}
        f:test/bar: {}
        f:test/foo: {}
      f:finalizers:
        .: {}
        v:"composite.apiextensions.crossplane.io": {}
      f:generateName: {}
      f:labels:
        .: {}
        f:bar: {}
        f:crossplane.io/claim-name: {}
        f:crossplane.io/claim-namespace: {}
        f:crossplane.io/composite: {}
        f:foo: {}
    f:spec:
      .: {}
      f:claimRef:
        .: {}
        f:apiVersion: {}
        f:kind: {}
        f:name: {}
        f:namespace: {}
      f:compositionRef:
        .: {}
        f:name: {}
      f:compositionRevisionRef:
        .: {}
        f:name: {}
      f:compositionUpdatePolicy: {}
      f:coolField: {}
      f:numbers: {}
      f:parameters:
        .: {}
        f:tags:
          .: {}
          f:newtag: {}
          f:tag: {}
      f:resourceRefs: {}
  manager: crossplane
  operation: Update
  time: "2023-11-23T12:16:07Z"
- apiVersion: nop.example.org/v1alpha1
  fieldsType: FieldsV1
  fieldsV1:
    f:status:
      .: {}
      f:conditions:
        .: {}
        k:{"type":"Ready"}:
          .: {}
          f:lastTransitionTime: {}
          f:reason: {}
          f:status: {}
          f:type: {}
        k:{"type":"Synced"}:
          .: {}
          f:lastTransitionTime: {}
          f:reason: {}
          f:status: {}
          f:type: {}
      f:coolerField: {}
  manager: crossplane
  operation: Update
  subresource: status
  time: "2023-11-23T12:16:07Z"
```

So, even we use patching applicator, the operation set by API server is `Update`. With introducing server-side apply, we are going to get another field manager listed here, with the operation set to `Apply`. Two field managers are going to share the same set of fields.

As the consequence, if we remove field/label/annotation on the existing claim, that change is NOT going to propagate down to the composite, i.e. the reported bugs are still valid and not fixed.

IMHO, Crossplane behavior should be the same both on existing and the new claims. How is user supposed to understand that label removals are propagated only on claims created by Crossplane version using server-side apply?

> But we need this for new claims as well since we use CREATE

Even if we stop using create, the issue is still valid for the existing claims, and `maybeFixFieldOwnership` takes care about that.

Reply:
> IMHO, Crossplane behavior should be the same both on existing and the new claims. How is user supposed to understand that label removals are propagated only on claims created by Crossplane version using server-side apply?

Agreed. 
As we discussed offline, I am okay with one-time migrations (and this code eventually being dropped after the previous release dropped from the supported releases). I am more concerned about this being part of regular operation, i.e. for all new composites, we always do a migration because it was created with a Create.

Reply:
As promised, I had a look at `client-go` [UpgradeManagedFields()](https://github.com/kubernetes/client-go/blob/master/util/csaupgrade/upgrade.go#L81) function, but unfortunately we cannot use it:

* it does not give a feedback if patching was performed, so cannot know if we need to send update to API server
* it does not support skipping subresources (important for patching composites)

Other than that, the patching logic in our function is the same as in `UpgradeManagedFields()`.